File: EPI/5.11/swift/5.11_test.swift
// swiftlint:disable variable_name

import Darwin

func == (lhs: Rect, rhs: Rect) -> Bool {
    return lhs.origin == rhs.origin && lhs.size == rhs.size
}

let tests: [(a: Rect, b: Rect, intersection: Rect?)] = [
    (
        Rect(origin: (0, 0), size: (5, 5)),
        Rect(origin: (1, 1), size: (2, 3)),
        Rect(origin: (1, 1), size: (2, 3))
    ),
    (
        Rect(origin: (0, 0), size: (5, 5)),
        Rect(origin: (-5, -5), size: (7, 8)),
        Rect(origin: (0, 0), size: (2, 3))
    ),
    (
        Rect(origin: (0, 0), size: (5, 5)),
        Rect(origin: (5, 4), size: (1, 7)),
        Rect(origin: (5, 4), size: (0, 1))
    ),
    (
        Rect(origin: (0, 0), size: (5, 5)),
        Rect(origin: (0, 6), size: (5, 5)),
        nil
    ),
]

for test in tests {
    for (a, b) in [(test.a, test.b), (test.b, test.a)] {
        let actual = intersection(a, b)
        guard actual == nil ? test.intersection == nil : actual! == test.intersection! else {
            print("For rectangles \(a) and \(b), " +
                "expected intersection to be \(test.intersection), but was \(actual)")
            exit(1)
        }
    }
}

File: EPI/Primitive Types/Rectangle intersection/typescript/5.11_test.ts
import {Rect, intersection} from './5.11';

declare function require(name: string): any;
const assert = require('assert');
const inspect = (v: any) => require('util').inspect(v, {depth: null});

const tests: {a: Rect, b: Rect, intersection: Rect | undefined}[] = [
    {
        a: {topLeft: {x: 0, y: 0}, bottomRight: {x: 5, y: 5}},
        b: {topLeft: {x: 1, y: 1}, bottomRight: {x: 3, y: 4}},
        intersection: {topLeft: {x: 1, y: 1}, bottomRight: {x: 3, y: 4}}
    },
    {
        a: {topLeft: {x: 0, y: 0}, bottomRight: {x: 5, y: 5}},
        b: {topLeft: {x: -5, y: -5}, bottomRight: {x: 2, y: 3}},
        intersection: {topLeft: {x: 0, y: 0}, bottomRight: {x: 2, y: 3}}
    },
    {
        a: {topLeft: {x: 0, y: 0}, bottomRight: {x: 5, y: 5}},
        b: {topLeft: {x: 5, y: 4}, bottomRight: {x: 6, y: 11}},
        intersection: {topLeft: {x: 5, y: 4}, bottomRight: {x: 5, y: 5}}
    },
    {
        a: {topLeft: {x: 0, y: 0}, bottomRight: {x: 5, y: 5}},
        b: {topLeft: {x: 0, y: 6}, bottomRight: {x: 5, y: 11}},
        intersection: undefined
    },
];

for (const test of tests) {
    const actual = intersection(test.a, test.b);
    assert.deepStrictEqual(actual, test.intersection,
        `For rectangles ${inspect(test.a)} and ${inspect(test.b)}, ` +
        `expected intersection to be ${inspect(test.intersection)}, but was ${inspect(actual)}`);
}
